documentation/testing.asciidoc

```diff
@@ -0,0 +1,208 @@
+          .when()
+          .get("/animals")
+          .then()
+          .statusCode(200)
+          .extract().response();
+
+        PageResultDTO<AnimalDTO> animalsReturned = response.as(new TypeRef<PageResultDTO<AnimalDTO>>() {
+        });
+
+        assertEquals(2, animalsReturned.getTotalElements());
+    }
+}
+----
+
+When run, tkit-test extension will first start all the docker containers we need, update Quarkus test configuration and then Quarkus will bootstrap our application.
+Our test invokes REST API of our app using RestAsurred client(remote HTTP call) so we are indeed testing real functionality of our app, including persistence and DI.
+Figure bellow illustrated the test execution sequence:
+
+
+.Testing with tkit-test & testcontainers
+image:/documentation/images/testing-setup.svg[
+"Testing Setup",width=512
+link="/documentation/images/testing-setup.svg"]
+
+== Testing in native mode
+
+The default test setup runs the tests agains a JVM build of our app. But as our test interact with the app via REST API, we can easily swap the app under test for a native build of our app - while reusing 100% of our test code. 
+
+In the docker-compose we use for test env bootstrap, we can add our app's native image and then we redirect our rest-assured calls to it. If you have multiple tests, and also use native tests, then we suggest a small refactoring of the test classes:
+
+.Native build of our app in docker-compose
+
+[source,yaml]
+----
+...
+my-native-app:
+  container_name: my-native-app
+  image: my-native-app:latest
+  ports:
+    - "8080:8080"
+  labels:
+    - "test.unit=false"
+    - "test.priority=200"
+    - "test.image.pull=DEFAULT"
+    - "test.Wait.forLogMessage.regex=.*Installed features:.*"
+    - "test.Wait.forLogMessage.times=1"
+    - "test.log=true"
+...
+----
+
+.Common abstract class for all tests
+
+[source,java]
+----
+@QuarkusTestResource(DockerComposeTestResource.class)
+public abstract class AbstractTest {
+
+    @DockerService("my-native-app")
+    protected DockerComposeService service;
+
+    @BeforeEach
+    public void before() {
+        # when we are in native mode, then we have `my-native-app` service, and we redirect all rest assured calls to it
+        RestAssured.enableLoggingOfRequestAndResponseIfValidationFails();
+        if (service != null) {
+            RestAssured.port = service.getPort(8080);
+            RestAssured.baseURI = "http://" + service.getHost();
+        }
+    }
+}
+
+@QuarkusTest
+public class MyRestControllerTest extends AbstractTest {
+    
+  @Test
+  @WithDBData(value = "data/animal.xls", deleteBeforeInsert = true)
+  public void getAllTest() {
+
+      Response response = given()
+        .when()
+        .get("/animals")
+        .then()
+        .statusCode(200)
+        .extract().response();
+
+      PageResultDTO<AnimalDTO> animalsReturned = response.as(new TypeRef<PageResultDTO<AnimalDTO>>() {
+      });
+
+      assertEquals(2, animalsReturned.getTotalElements());
```

Comment:
From my experience I see two kind of REST tests:
1. regression testing to ensure the contract is not changed by accident. For that I would use things like https://github.com/approvals/approvaltests.java
2. functional testing of the service. Here I would simply use the REST client and avoid manually fiddling with JSON mapping, etc.

Just my personal point of view and my five cents...